enhancements/coreos-bootimages.md

```diff
@@ -0,0 +1,143 @@
+
+This enhancement calls for a standardized JSON format for (RHEL) CoreOS bootimage metadata to be placed at https://mirror.openshift.com *and* included in a new `rhel-coreos-bootimages` image included in the release image.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Since the initial release of OpenShift 4, we have "pinned" RHCOS bootimage metadata inside [openshift/installer](https://github.com/openshift/installer).
+In combination with the binding between the installer and release image, this means that everything needed to install OpenShift (including the operating system "bootimages" such as e.g. AMIs and OpenStack `.qcow2` files) are all captured behind the release image which we can test and ship as an atomic unit.
+
+We have a mechanism to do [in place updates](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md), but there is no automated mechanism to update "bootimages" past a cluster installation.
+
+This enhancement does not describe an automated mechanism to do this: the initial goal is to include this metadata in a standardized format in the cluster and at mirror.openshift.com so that UPI installations can do this manually, and we can start work on an IPI mechanism.
+
+### Stream metadata format
+
+As part of unifying Fedora CoreOS and RHEL CoreOS, we have standardized on the "stream metadata" format used by FCOS.  More in [FCOS docs](https://docs.fedoraproject.org/en-US/fedora-coreos/getting-started/) and [this RHCOS issue](https://github.com/openshift/os/issues/477).
+
+There is a new [stream-metadata-go](https://github.com/coreos/stream-metadata-go) library to consume this data.
+
+### Adding a new git repository and release image component + configmap with this data
+
+A new git repository https://github.com/openshift/rhel-coreos-bootimages will be created (based on [an existing prototype repository](https://github.com/cgwalters/rhel-coreos-bootimages)).  It will have the stream JSON.
+
+Additionally, this repository will be included in the release image as `rhel-coreos-bootimages` and it will use the CVO to install a `configmap/coreos-bootimages` in the `openshift-machine-config-operator` namespace.
+
+### Update `oc adm release new` to patch the installer binary to inject this data
+
+The installer needs CoreOS to start at all; today the bootimages are used for the bootstrap node
+and control plane which are created by the installer directly.  In order to avoid duplicating the
+metadata, change `oc adm release new` to *binary patch* the included `openshift-install`
```

Comment:
Will the binary patch inject the entire stream metadata into the installer? What is the process that a developer (or a libvirt user) running a locally built installer would use? Would they be expected to do a binary patch as well?

Reply:
Today the installer when built from git master defaults to an OKD release image with `rhcos.json`, which just doesn't work at all - this is a common stumbling block for people.

I think probably the simplest thing here is that when building from git branch `master`, we default to using `oc` to binary patch to the latest release image data.

Reply:
I am confused about the statement that it "just doesn't work at all". I frequently do successful installation using the installer built from git master without needing to make any changes for the RHCOS image.

Reply:
When you run your locally-built installer, do you provide `env OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE`?

What I and others have hit is pretty simple:

The [default release image](https://github.com/openshift/installer/blob/60810de099458d4ad57d0999b7ae2c38440473b8/pkg/asset/releaseimage/default.go#L23) is OKD:

```
$ oc image info registry.ci.openshift.org/origin/release:4.8 | grep -i release=
Labels:      io.openshift.release=4.8.0-0.okd-2021-03-07-094455
$
```

Yet the [default pinned bootimages](https://github.com/openshift/installer/blob/master/data/data/rhcos.json) are RHCOS.  That means e.g. the bootstrap node is RHCOS yet we're pulling down the OKD release image which causes various problems, at least in the past.

We added support to [streamline overriding the release image](https://github.com/cgwalters/xokdinst/pull/4) to xokdinst to help with this for example.